graphix/simulator.py

```diff
@@ -6,29 +6,87 @@
 
 from __future__ import annotations
 
+import abc
 import warnings
 
 import numpy as np
 
-from graphix.command import CommandKind
+import graphix.clifford
+from graphix.command import BaseM, CommandKind, M
+from graphix.pauli import MeasureUpdate
+from graphix.sim.base_backend import Backend, MeasurementDescription
 from graphix.sim.density_matrix import DensityMatrixBackend
 from graphix.sim.statevec import StatevectorBackend
 from graphix.sim.tensornet import TensorNetworkBackend
 
 
+class MeasureMethod(abc.ABC):
+    """
+    Simulators are now parametrized by MeasureMethod, with default measurement method that implements MBQC.
+    It allows customized measurement methods, for instance for delegated QC protocols.
+    """
+
+    def measure(self, backend: Backend, cmd, noise_model=None) -> None:
+        description = self.get_measurement_description(cmd)
+        result = backend.measure(cmd.node, description)
+        if noise_model is not None:
+            result = noise_model.confuse_result(result)
+        self.set_measure_result(cmd.node, result)
+
+    @abc.abstractmethod
+    def get_measurement_description(self, cmd: BaseM) -> MeasurementDescription: ...
+
+    @abc.abstractmethod
+    def get_measure_result(self, node: int) -> bool: ...
+
+    @abc.abstractmethod
+    def set_measure_result(self, node: int, result: bool) -> None: ...
+
+
+class DefaultMeasureMethod(MeasureMethod):
+    """
+    Default measurement method implementing standard measurement plane/angle update for MBQC.
+    To be overwritten by custom measurement methods in the case of delegated QC protocols.
+    """
+
+    def __init__(self, results=None):
+        if results is None:
+            results = dict()
+        self.results = results
+
+    def get_measurement_description(self, cmd: BaseM) -> MeasurementDescription:
+        assert isinstance(cmd, M)
+        angle = cmd.angle * np.pi
+        # extract signals for adaptive angle
+        s_signal = sum(self.results[j] for j in cmd.s_domain)
+        t_signal = sum(self.results[j] for j in cmd.t_domain)
+        measure_update = MeasureUpdate.compute(cmd.plane, s_signal % 2 == 1, t_signal % 2 == 1, graphix.clifford.I)
+        angle = angle * measure_update.coeff + measure_update.add_term
+        return MeasurementDescription(measure_update.new_plane, angle)
+
+    def get_measure_result(self, node: int) -> bool:
+        return self.results[node]
+
+    def set_measure_result(self, node: int, result: bool) -> None:
+        self.results[node] = result
+
+
 class PatternSimulator:
     """MBQC simulator
 
     Executes the measurement pattern.
     """
 
-    def __init__(self, pattern, backend="statevector", noise_model=None, **kwargs):
+    def __init__(
+        self, pattern, backend="statevector", measure_method: MeasureMethod = None, noise_model=None, **kwargs
```

Comment:
You need to add `| None` to the argument type when you pass `None` as default value.